File: dmoj/gfssoc16s2/main.py
def ace(n):
    lookup = [10, 12, 14]
    base = 1
    i = 0

    while n >= 0:
        i += lookup[n % 3] * base
        n = n // 3 - 1
        base *= 16

    return i


def lower_bound(x):
    lo, hi = 0, 1

    while ace(hi) < x:
        hi *= 2

    while lo < hi:
        mid = (lo + hi) // 2

        if ace(mid) < x:
            lo = mid + 1
        else:
            hi = mid

    return lo


def upper_bound(x):
    lo, hi = 0, 1

    while not x < ace(hi):
        hi *= 2

    while lo < hi:
        mid = (lo + hi) // 2

        if not x < ace(mid):
            lo = mid + 1
        else:
            hi = mid

    return lo


a = int(input())
b = int(input())

print(upper_bound(b) - lower_bound(a))

File: dmoj/gfssoc16s2/main.cpp
#include <iostream>
#define pscan(x) do{while((x=getchar())<'0'); for(x-='0'; '0'<=(_=getchar()); x=(x<<3)+(x<<1)+_-'0');}while(0)
char _;

inline long long int ace(int n) {
    static const int lookup[3]{10, 12, 14};
    long long int base = 1, i = 0;

    while (n >= 0) {
        std::lldiv_t div = std::lldiv(n, 3);
        i += lookup[div.rem] * base;
        n = div.quot - 1;
        base <<= 4;
    }

    return i;
}

inline int lowerBound(long long int x) {
    int lo = 0, hi = 1;

    while (ace(hi) < x) hi <<= 1;

    while (lo < hi) {
        int mid = (lo + hi) >> 1;

        if (ace(mid) < x)
            lo = mid + 1;
        else
            hi = mid;
    }

    return lo;
}

inline int upperBound(long long int x) {
    int lo = 0, hi = 1;

    while (ace(hi) <= x) hi <<= 1;

    while (lo < hi) {
        int mid = (lo + hi) >> 1;

        if (ace(mid) <= x)
            lo = mid + 1;
        else
            hi = mid;
    }

    return lo;
}

long long int a, b;

int main() {
    pscan(a);
    pscan(b);

    printf("%d", upperBound(b) - lowerBound(a));

    return 0;
}